### PR TITLE
Dispensary flow

### PR DESCRIPTION
**Pull request by Shalashtein**
- Typo fixed across different files and database
- Data Types fixed on different attributes in database
- Added annotate gem and run rake task
- Added AASM state management gem, and used it to run validation when a transfer of material is initiated
- Edited seed file to reflect changes

### CODE REVIEW

File: app/models/material.rb [hunk 1]
```diff
@@ -1,21 +1,181 @@
...
-    self.save
-    Transfer.create(sender_material: self,
-                    reciever_material: reciever_material,
-                    sender: self.owner,
-                    reciever: reciever,
-                    weight: amount)
-    reciever_material
+  belongs_to :owner, class_name: "User", foreign_key: 'owner_id'
+  has_many :transfers, class_name: "Transfer", foreign_key: 'sender_material_id', dependent: :destroy
+  has_many :transfers, class_name: "Transfer", foreign_key: 'receiver_material_id', dependent: :destroy
```
Comment (wolak88) [inline app/models/material.rb:88]: We can't have two relations with the same name, maybe let's go with `outgoing_transfers` and `incoming_transfers`

File: app/models/material.rb [hunk 2]
```diff
@@ -1,21 +1,181 @@
...
+  # Returns an array, the first element will be a boolean that describes
+  # whether or not the tranfer was successful. If the tranfer is successful
+  # it will be passed as the second element, if no the second element would
+  # be the state of the Material to describe where it's stuck
+  def split(receiver, amount)
+
+    # Run validations
+    process(receiver, amount)
+
+    # Process tranfer if all states transitioned
```
Comment (wolak88) [inline app/models/material.rb:100]: nit: `transfer`

File: db/seeds.rb [hunk 3]
```diff
@@ -1,28 +1,24 @@
...
-MaterialType.delete_all
-Material.delete_all
-Transfer.delete_all
+puts 'Destroying old records...'
+Transfer.destroy_all
+MaterialType.destroy_all
+User.destroy_all
+Material.destroy_all
+Profile.destroy_all
+Address.destroy_all
```
Comment (wolak88) [inline db/seeds.rb:9]: Why are we not removing Pages, Posts, Dispensaries?

File: app/models/material.rb [hunk 4]
```diff
@@ -1,21 +1,181 @@
...
 class Material < ApplicationRecord
+  include AASM
+
+  # No direct assignment ensures that the state of the model can only be modified
+  # by going through the transitions abd guards speicifed here, this way we might
+  # prevent problematic hacks elsewhere in the code.
+  #
+  #
+
+  aasm no_direct_assignment: true do
```
Comment (wolak88) [inline app/models/material.rb:41]: Let's set `whiny_persistence: true` and we will use all transitions with bang `!`

File: app/models/material.rb [hunk 5]
```diff
@@ -1,21 +1,181 @@
...
+
+  def process(receiver, amount)
+    start_validation
+    validate_dispensary
+    validate_stock(amount)
+    validate_receiver(receiver)
+    validate_evaluation(receiver)
+    validate_amount_allowed(receiver, amount)
+    validate_credits(receiver, amount)
+  end
```
Comment (wolak88) [inline app/models/material.rb:146]: I'm not sure about this approach, what if one of the intermediate steps fail? 
I think we will have sender_material which is locked and we don't have any way to put it back in ready state.

We can take advantage of some hooks(example code from my project):
```
event Journeys::CommonEventEnum::SUBMIT_FOR_CONSULTATION.to_sym do
          # eCommerce in state, LTDD
          transitions(
            from: Journeys::CommonStatusEnum::QUIZ_PENDING.to_sym,
            to: Journeys::CommonStatusEnum::CONSULTATION_PENDING.to_sym,

            guard: -> { user_completed_quiz?(event: Journeys::CommonEventEnum::SUBMIT_FOR_CONSULTATION.to_sym) }
          )

          # eCommerce out state
          transitions(
            from: Journeys::CommonStatusEnum::QUIZ_PENDING.to_sym,
            to: Journeys::CommonStatusEnum::PRESCRIPTION_TRANSFER_PENDING.to_sym,

            guard: -> { e_commerce_rx_out_of_state?(event: Journeys::CommonEventEnum::SUBMIT_FOR_CONSULTATION.to_sym) },

            after: lambda {
              update!(doctor_products: user_products)
            }
          )

          after lambda {
            update_consultation! unless user.out_of_state?
          }

          after_commit lambda {
            Analytics.track(
              user_id: user.id,
              event: 'ConsultationQuizFinished'
            )
            UserEvent.send(user: user, event: UserEventEnum::QUIZ_FINISHED)

            sync_to_kmg! unless user.out_of_state?
          }
        end
```

**Review by wolak88 — CHANGES_REQUESTED** [review]

File: db/seeds.rb (hunk 3 above)
Comment (Shalashtein) [reply db/seeds.rb:9]: We aren't using them yet, so far in our app flow and in the seeds we aren't creating Pages or Posts or Dispensaries

File: app/models/material.rb (hunk 1 above)
Comment (Shalashtein) [reply app/models/material.rb:88]: done

File: app/models/material.rb (hunk 4 above)
Comment (Shalashtein) [reply app/models/material.rb:41]: done

File: app/models/material.rb (hunk 5 above)
Comment (Shalashtein) [reply app/models/material.rb:146]: I tried, let's see

File: app/models/material.rb [hunk 6]
```diff
@@ -1,23 +1,204 @@
...
+                               weight: amount)
+    [true, transfer]
+  rescue AASM::InvalidTransition, ArgumentError => e
+    [false, error_message(e), e]
+  end
+
+  private
+
+  def validate_transfer(receiver, amount)
+    start_tranfer
```
Comment (wolak88) [inline app/models/material.rb:136]: nit: `transfer`

**Review by wolak88 — APPROVED** [review]

**Pull request merged**